docs: add guide for operation complexity controls (#4402)
graphql/graphql-js: website/pages/docs/operation-complexity-controls.mdx

---
title: Operation Complexity Controls
---

# Operation Complexity Controls

GraphQL gives clients a lot of flexibility to shape responses, but that
flexibility can also introduce risk. Clients can request deeply nested fields or 
large volumes of data in a single operation. Without controls, these operations can slow
down your server or expose security vulnerabilities.

This guide explains how to measure and limit operation complexity in GraphQL.js
using static analysis. You'll learn how to estimate the cost
of an operation before execution and reject it if it exceeds a safe limit.

<Callout type="info" emoji="ℹ️">

>   In production, we recommend using [trusted documents](https://graphql.org/learn/persistence/)

Reviewer: I think this link was hallucinated, but we can link to #4418 now instead?

This URL is a guess; we should check it:

```suggestion
  In production, we recommend using [trusted documents](/docs/going-to-production#only-allow-trusted-documents)
```

  rather than analyzing arbitrary documents at runtime. Complexity analysis can still be
  useful at build time to catch expensive operations before they're deployed.
</Callout>

## Why complexity control matters

GraphQL lets clients choose exactly what data they want. That flexibility is powerful, 
but it also makes it hard to predict the runtime cost of a query just by looking
at the schema.

Without safeguards, clients could:

- Request deeply nested object relationships

> - Use recursive fragments to multiply field resolution

Reviewer: GraphQL validation has always explicitly forbid recursive fragments ([fragment spreads must not form cycles](https://spec.graphql.org/July2015/#sec-Fragment-spreads-must-not-form-cycles)) so servers that allow this are not GraphQL compliant. Let's be more precise what we mean here:

```suggestion
- Use nested fragments to multiply field resolution
```

import { schema } from './schema.js';

const source = `
  query {
    users {
      id
      posts {
        title
      }
    }
  }
`;

const document = parse(source);

const result = await graphql({
  schema,
  source,
  validationRules: [
    ...specifiedRules,
    createComplexityRule({
      estimators: [simpleEstimator({ defaultComplexity: 1 })],
      maximumComplexity: 50,
      onComplete: (complexity) => {
        console.log('Query complexity:', complexity);
      },
    }),
  ],
});

console.log(result);
```

<Callout type="info" emoji="ℹ️">
  Only use complexity rules in validation if you're sure the analysis is cycle-safe.
  Otherwise, run complexity checks after validation and before execution.
</Callout>

## Complexity in trusted environments

In environments that use persisted or precompiled operations, complexity analysis is still
useful, just in a different way. You can run it at build time to:

- Warn engineers about expensive operations during development
- Track changes to operation cost across schema changes
- Define internal usage budgets by team, client, or role

## Best practices

> - Use trusted documents in production when possible.

Reviewer: ```suggestion
- Only accept trusted documents in production when possible.
```

> - Use complexity analysis as a development-time safeguards.

Reviewer: ```suggestion
- Use complexity analysis as a development-time safeguard.
```
